urlc8dbc7ce6e_dlubarov_guava_tgz-pJ8-MainJ8/src/d/BytecodeMethodDef.java

```diff
@@ -0,0 +1,467 @@
+
+                    case PUT_STATIC_FIELD: {
+                        BaseObject newValue = stack[sp--];
+                        int typeTableIndex = code[ip++];
+                        int staticFieldIndex = code[ip++];
+                        rawTypeTable[typeTableIndex].staticFields[staticFieldIndex] = newValue;
+                        break;
+                    }
+
+                    case GET_INSTANCE_FIELD: {
+                        int stringTableIndex = code[ip++];
+                        String fieldName = stringTable[stringTableIndex];
+                        assert fieldName.intern() == fieldName : "String in literal pool wasn't interned.";
+
+                        BaseObject target = stack[sp--];
+                        int fieldIndex = target.type.rawType.virtualFieldTable.get(fieldName);
+                        BaseObject value = target.fields[fieldIndex];
+                        assert value != null : String.format(
+                                "%s's instance field #%d is null.",
+                                target, fieldIndex);
+                        stack[++sp] = value;
+                        break;
+                    }
+
+                    case PUT_INSTANCE_FIELD: {
+                        int stringTableIndex = code[ip++];
+                        String fieldName = stringTable[stringTableIndex];
+                        assert fieldName.intern() == fieldName : "String in literal pool wasn't interned.";
+
+                        BaseObject target = stack[sp--];
+                        BaseObject newValue = stack[sp--];
+                        assert target.type.rawType.virtualFieldTable.containsKey(fieldName) :
+                            String.format("%s does not have field %s", target, fieldName);
+                        int fieldIndex = target.type.rawType.virtualFieldTable.get(fieldName);
+                        target.fields[fieldIndex] = newValue;
+                        break;
+                    }
+
+                    case INVOKE_STATIC: {
+                        int methodTableIndex = code[ip++];
+                        ConcreteMethodDef m = (ConcreteMethodDef) methodTable[methodTableIndex];
+                        BaseObject a = desc.isStatic ? null : stack[bp + 1]; // current object
+
+                        // Create array of generic arguments.
+                        int numGenericArgs = m.desc.numGenericParams;
+                        ConcreteType[] newGenericArgs;
+                        if (numGenericArgs == 0)
+                            // We don't want to allocate an empty array, that's wasteful.
+                            newGenericArgs = null;
+                        else {
+                            newGenericArgs = new ConcreteType[numGenericArgs];
+                            for (int j = 0; j < numGenericArgs; ++j)
+                                newGenericArgs[j] = fullTypeTable[code[ip++]].toConcrete(a, genericArgs);
+                        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 206:213, 233:240.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.